pkg/sql/catalog/collectionfactory/collection_factory.go

```diff
@@ -0,0 +1,17 @@
+// Copyright 2022 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package collectionfactory
+
+// CollectionFactory is an interface for *descs.CollectionFactory.
+// It's created to avoid cyclic dependency.
+type CollectionFactory interface {
+	ImplmentCollectionFactory()
+}
```

Comment:
this smells to me. What is the circularity? Can we define narrower interfaces at the relevant call sites?

Reply:
It's the loop between`cloud` and `descs`.

```
.-> //pkg/cloud:cloud (d5981f1a3d4a46828dfe84289c55e9bfde67ac814b3f9fd2987adc5d9089163d)
|   //pkg/sql/catalog/descs:descs (d5981f1a3d4a46828dfe84289c55e9bfde67ac814b3f9fd2987adc5d9089163d)
|   //pkg/sql/catalog/hydrateddesc:hydrateddesc (d5981f1a3d4a46828dfe84289c55e9bfde67ac814b3f9fd2987adc5d9089163d)
|   //pkg/sql/catalog/lease:lease (d5981f1a3d4a46828dfe84289c55e9bfde67ac814b3f9fd2987adc5d9089163d)
|   //pkg/multitenant:multitenant (d5981f1a3d4a46828dfe84289c55e9bfde67ac814b3f9fd2987adc5d9089163d)
`-- //pkg/cloud:cloud (d5981f1a3d4a46828dfe84289c55e9bfde67ac814b3f9fd2987adc5d9089163d)
```

I encountered this because I wanted to have a `CollectionFactory` field in `ExternalStorageContext`, so that we can pass it to `InternalFileToTableExecutor` and use it in `file_table_read_write.go`.

Reply:
> Can we define narrower interfaces at the relevant call sites?

I'm not sure. In this PR we only refactored the use cases with DDLs. In the future we will want to migrate [all usages](https://docs.google.com/spreadsheets/d/1MQV5IE-01_Owfrp9HIIfXIIERXyTjvzA8oa0SUNBysU/edit?usp=sharing) with not-nil txn to `CollectionFactory.TxnWithExecutor()`. Not sure if we will encounter a larger number and more complex dependency loops.

Reply:
> I wanted to have a CollectionFactory field in ExternalStorageContext, so that we can pass it to InternalFileToTableExecutor

For more context, I was following how `InternalFileToTableExecutor.ie` gets written to construct the same path for `InternalFileToTableExecutor.cf` .

Reply:
When a dependency cycle gets in the way, the answer is almost never to smuggle an object through an interface. For a hack, for a brief moment, it can be okay. In some cases, where the idea of opaque dependency injection is deeply fundamental to the problem, then it can be the right thing. In general, it's a code smell. [This PR](https://github.com/cockroachdb/cockroach/pull/87140) should fix the dependency cycle.

Reply:
I see, thanks for explaining! I just rebased from that PR and it nicely solved the loop. Will think twice when I encounter a dependency loop next time!